### PR TITLE
Add BerryLM

### PR DESCRIPTION
**Pull request by DanilSmorchkov**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48901) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48901)

# What does this PR do?

Adds **BerryLM**, a hybrid linear/full-attention mixture-of-experts decoder with a thinking mode
(`BerryLMConfig`, `BerryLMModel`, `BerryLMForCausalLM`), released as `RWB/BerryLM-OS` (~2.8B active / ~18B
total parameters, 40 layers, 180k text-only vocabulary, 256k context) with a tiny random CI checkpoint
`RWB/BerryLM-OS-tiny-random`.

Architecture (details in `docs/source/en/model_doc/berrylm.md`):

- **Hybrid attention**: three linear-attention layers per full-attention layer. The linear layers run a gated delta
  rule with a **per-channel forget gate** (a rank-128 projection of the layer input sets a separate log-decay for every
  key channel of every head; the Kimi Delta Attention recurrence), the full-attention layers are GQA with a sigmoid
  output gate and partial rotary embeddings.
- **Gated Block AttnRes**: each decoder layer reads a softmax mixture of the residual streams committed at block
  boundaries (embeddings = block 0) instead of the plain residual stream, gated toward the identity by a per-layer
  scalar (`y = x + tanh(g) * (mix - x)`; `g = 0` is the exact identity). Token-local, transparent to the caches.
- **Sparse MoE**: 128 routed experts (top-8, renormalized softmax) plus a sigmoid-gated shared expert.

Implementation notes:

- Standalone modeling (no `modular_*` file): the KDA layer and the AttnRes mixer have no upstream equivalent, and the
  other blocks are small; kernels come through `use_kernel_func_from_hub_with_fallback` (`flash-linear-attention`
  `chunk_kda` / `fused_recurrent_kda`, `causal-conv1d`) with pure-torch references as fallbacks, so the model runs on
  CPU / without the optional packages.
- `@strict` dataclass config with typed fields and `__post_init__` validation (layer types, head divisibility, block
  size, safe-gate bound).
- `_can_compile_fullgraph = True`: the single-token KDA step runs the torch reference under `torch.compile`
  (flash-linear-attention's recurrent KDA kernel is `torch.compiler.disable`d), so the static-cache / compiled
  generation tests pass with or without the optional kernels.
- Caching through the generic `DynamicCache` linear-attention layer (conv + recurrent state), padding masked on
  continued forwards; gradient checkpointing safe (the block streams are proper checkpoint inputs).
- Auto mappings (`AutoConfig`, `AutoModel`, `AutoModelForCausalLM`), TP/PP/EP plans in the config.

Fixes # (n/a — new model)

## Tests

`tests/models/berrylm/test_modeling_berrylm.py`: the `CausalLMModelTester` suite plus model-specific tests — chunked
vs recurrent KDA equivalence, AttnRes zero-gate identity (block size 0 == block size N with zero gates), cached
multi-token vs single-token decode, recurrent layers masking padding on continued forwards, padding-free
(`position_ids`-only) forward, device-map loading.

```bash
python -m pytest tests/models/berrylm -v                      # 153 passed, 163 skipped on 1 GPU (fla + causal-conv1d installed)
python -m pytest tests/models/berrylm -v -k test_training_overfit   # CPU, torch fallbacks (no fla)
python utils/checkers.py "sort_auto_mappings, init_isort, auto_mappings, imports, import_complexity, config_attributes, \
  config_docstrings, docstrings, inits, repo, modeling_structure, doc_toc, dummies, noisy_comments, copies"
ruff check src/transformers/models/berrylm tests/models/berrylm && ruff format --check src/transformers/models/berrylm tests/models/berrylm
```

Numerics against the training implementation and the vLLM / SGLang ports of the same model are in the release
model card (`RWB/BerryLM-OS`).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the contributor guideline and the Pull Request checks?
- [ ] Was this discussed/approved via a Github issue or the forum?
- [x] Did you make sure to update the documentation with your changes according to the guidelines?
- [x] Did you write any new necessary tests?

## Who can review?

Text models: @ArthurZucker @Cyrilvallez @vasqu

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hey, I'm sorry, we generally don't add new models to the core library until there are significant pre-trained models with that architecture! Demo models don't really count, and I can't find the mentioned checkpoints like `RWB/BerryLM-OS` on the Hub. Maybe they're private?

**Comment by DanilSmorchkov** [comment]
Thanks for taking a look, and sorry for the confusion! This isn't a demo architecture: the checkpoint
is a fully trained 18B-parameter hybrid MoE (AttnRes + KDA linear attention, the architecture in this PR),
the weights are final, and we're finishing the release (model card, licensing, engine support —
the matching vLLM and SGLang PRs are open in parallel: vllm-project/vllm#57384, sgl-project/sglang#39972).

The Hub repo `RWB/BerryLM-OS` is currently private; we plan to make it public within the next two weeks,
together with the modeling code as remote code, so the checkpoint is usable right away and this PR
upstreams the same implementation.

Happy to convert this PR to a draft until the weights are public — I'll ping you here as soon as the repo
is live.

**Comment by Rocketknight1** [comment]
Got it! We'll probably wait to see how the remote code variant does. We usually only add models to the core library when there's a significant amount of users, because the Transformers team takes over maintenance at that point. Let us know if there are any issues with the remote code port until then, though!
